### PR TITLE
secp256k1: Faster scalar mult via width-5 wNAF.

### PR DESCRIPTION
**Pull request by davecgh**
**This requires #3731**.

This improves the performance of the non-constant-time scalar multiplication implementation via width-5 windowed non-adjacent form (wNAF) recoding.

It introduces an optimized width-5 recoder together with precomputed odd-multiple tables for both P and φ(P) which allows the existing endomorphism-based simultaneous multiplication to replace the previous width-2 NAF representation with the sparser width-5 wNAF representation.

On average, width-5 wNAF reduces the density of non-zero digits from approximately one in three to one in six which significantly reduces the number of point additions required during scalar multiplication at the cost of a small amount of additional precomputation.
    
Comprehensive tests are included to verify the mathematical properties of the resulting encodings, reconstruct the original values, and cover both deterministic edge cases and randomized balanced scalars. A benchmark for the recoder is also added.

Finally, it removes the old ordinary width-2 NAF code and associated tests and benchmark since it is no longer used.

It is implemented via a series of commits to ease the review process. See the individual commit descriptions for more details.

The following benchmark shows a before and after comparison:

```
name                 old time/op   new time/op   delta
---------------------------------------------------------------------------
ScalarMultNonConst   94.4µs ± 1%   84.3µs ± 1%   -10.75%  (p=0.000 n=10+10)
```

### CODE REVIEW

File: dcrec/secp256k1/curve.go [hunk 1]
```diff
@@ -1307,6 +1307,111 @@ func naf(k []byte) nafScalar {
...
+
+	// Build the corresponding odd multiples of φ(P).
+	//
+	// As above, both positive and negative entries are ultimately needed.  The
+	// sign of this table is chosen independently from the P table because it
+	// depends only on the sign of k2.
+	//
+	// Note that φ is a group homomorphism, so:
+	//
+	//   φ(n*P) = n*φ(P),
```
Comment (davecgh) [inline dcrec/secp256k1/curve.go]:
```suggestion
	//   φ(n*P) = n*φ(P)
```

File: dcrec/secp256k1/curve.go [hunk 2]
```diff
@@ -1384,94 +1470,91 @@ func ScalarMultNonConst(k *ModNScalar, point, result *JacobianPoint) {
...
-		if i >= m-k1Len {
-			k1BytePos, k1ByteNeg = k1PosNAF[i-(m-k1Len)], k1NegNAF[i-(m-k1Len)]
-		}
-		if i >= m-k2Len {
-			k2BytePos, k2ByteNeg = k2PosNAF[i-(m-k2Len)], k2NegNAF[i-(m-k2Len)]
+	for bit := maxBits; bit > 0; bit-- {
+		// Q = 2 * Q
+		DoubleNonConst(&q, &q)
+
+		// The encoded wNAF digit serves directly as an index into the
```
Comment (davecgh) [inline dcrec/secp256k1/curve.go]:
```suggestion
		// The encoded wNAF digit of k1 serves directly as an index into the
```

File: dcrec/secp256k1/curve.go [hunk 3]
```diff
@@ -1384,94 +1470,91 @@ func ScalarMultNonConst(k *ModNScalar, point, result *JacobianPoint) {
...
-			//
-			// +1: Q = Q + p2
-			// -1: Q = Q - p2
-			//  0: Q = Q (no change)
-			if k2BytePos&mask == mask {
-				AddNonConst(&q, p2, &q)
-			} else if k2ByteNeg&mask == mask {
-				AddNonConst(&q, p2Neg, &q)
-			}
+		// Likewise, the encoded wNAF digit serves directly as an index into the
```
Comment (davecgh) [inline dcrec/secp256k1/curve.go]:
```suggestion
		// Likewise, the encoded wNAF digit of k2 serves directly as an index into the
```

File: dcrec/secp256k1/curve.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Removed.

File: dcrec/secp256k1/curve.go (hunk 2 above)
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Added.

File: dcrec/secp256k1/curve.go (hunk 3 above)
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Updated.

**Review by jholdstock — APPROVED** [review]
